language/move-prover/bytecode-to-boogie/src/bytecode_instrs.bpl

```diff
@@ -15,16 +15,67 @@ function {:constructor} Path(p: [int]Edge, size: int): Path;
+
 type {:datatype} Value;
 function {:constructor} Boolean(b: bool): Value;
 function {:constructor} Integer(i: int): Value;
 function {:constructor} Address(a: Address): Value;
 function {:constructor} ByteArray(b: ByteArray): Value;
 function {:constructor} Str(a: String): Value;
-function {:constructor} Map(m: [Edge]Value): Value;
-function {:constructor} Vector(v: [Edge]Value, l: int): Value;
+// Currently empty vectors of arbitrary type will be identified as equal. If this causes a problem,
+// we need to add the element's TypeValue to the vector.
+function {:constructor} Vector(v: [int]Value, l: int): Value;
+// Currently structural equal structs are identified. We need to add the type name below if this causes a problem.
+function {:constructor} Struct(v: [FieldName]Value): Value;
+
+// Axiomatic definition of type test and equality.
+function has_type(t: TypeValue, v: Value): bool;
+function is_equal(t: TypeValue, v1: Value, v2: Value): bool;
+
+procedure {:inline 1} is_equal_proc(t: TypeValue, v1: Value, v2: Value) returns (r: Value);
+  ensures r == Boolean(is_equal(t, v1, v2));
+
+procedure {:inline 1} is_not_equal_proc(t: TypeValue, v1: Value, v2: Value) returns (r: Value);
+  ensures r == Boolean(!is_equal(t, v1, v2));
+
+axiom (forall x: bool :: has_type(BooleanType(), Boolean(x)));
+axiom (forall x: int :: has_type(IntegerType(), Integer(x)));
+axiom (forall x: Address :: has_type(AddressType(), Address(x)));
+axiom (forall x: ByteArray :: has_type(ByteArrayType(), ByteArray(x)));
+axiom (forall x: String :: has_type(StrType(), Str(x)));
+axiom (forall x: [int]Value, n: int, t: TypeValue :: has_type(VectorType(t), Vector(x, n)) <==>
+         (forall i: int :: i >= 0 && i < n ==> has_type(t, x[i])));
+axiom (forall name: TypeName, arity: int, ts: [FieldName]TypeValue, v: [FieldName]Value ::
+         has_type(StructType(name, arity, ts), Struct(v)) <==>
+           (forall f: FieldName :: f >= 0 && f < arity ==> has_type(ts[f], v[f])));
+
+axiom (forall x, y: bool :: is_equal(BooleanType(), Boolean(x), Boolean(y)) <==> x == y);
+axiom (forall x, y: int :: is_equal(IntegerType(), Integer(x), Integer(x)) <==> x == y);
+axiom (forall x, y: Address :: is_equal(AddressType(), Address(x), Address(y)) <==> x == y);
+axiom (forall x, y: ByteArray :: is_equal(ByteArrayType(), ByteArray(x), ByteArray(y)) <==> x == y);
+axiom (forall x, y: String :: is_equal(StrType(), Str(x), Str(y)) <==> x == y);
+axiom (forall t: TypeValue, x, y: [int]Value, n, m: int ::
+              is_equal(VectorType(t), Vector(x, n), Vector(y, m)) <==>
+                  n == m && (forall i: int :: i >= 0 && i < n ==> is_equal(t, x[i], y[i])));
+axiom (forall name: TypeName, arity: int, ts: [FieldName]TypeValue, v1, v2: [FieldName]Value ::
+         is_equal(StructType(name, arity, ts), Struct(v1), Struct(v2)) <==>
+           (forall i: int :: i >= 0 && i < arity ==> is_equal(ts[i], v1[i], v2[i])));
+
```

Comment:
The axiomatization of is_equal is nice.  I especially like the fact that is_equal(t, u, v) is left unconstrained if the type t does not match the structure of u and v.

But there is a problem also.  SMT solver handle universally-quantified facts by instantiating the bound variables.  The instantiation is typically done using a technique called e-matching (described nicely in Greg Nelson's thesis but perhaps there are more modern papers describing it) which requires triggers.  At the moment, I can't imagine what the triggers for the axioms above would be.  Have you tried to write down simple examples and prove equality/disequality assertions on them?  That could give us some clue about the completeness of this encoding.

Reply:
The funny thing is that this is really just a functional program with pattern matching. The "trigger" should be other side of L <==> R. I was thinking we tweak it as we go so Z3 can deal with it.